Refactored LinkedList Implementation: Encapsulation, Logging Fixes, and Comprehensive Testing

SeanSweeney04 opened the pull request:
Created a basic Linked List data structure using C++. It allows you to insert nodes, delete nodes and clear all the elements in the list. This task also focused on:

Implementing a flexible logging system
Exploring object creation and destruction
Understanding parameter passing and how it affects object lifetimes
Maintaining clean, SRP-compliant architecture throughout

Commit:
Initial commit

Commit:
Pull request 1

Commit:
Pull request 1

Commit:
Moved LinkedList2 into LinkedList and cleaned up repo

Commit:
fixed encapsulation in Node class

Commit:
fixed encapsulation in Node class

Commit:
restored the structure

Commit:
Renamed gitignore to .gitignore

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,156 @@
+#include <iostream>
```
Change header order: libs first, your includes next

SeanSweeney04 replied:
Changed

gregory-sokolov replied:
Not fixed, the order is still incorrect.

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,156 @@
+using namespace std;
+
+// Node class constructor: Initializes a node with given data and sets next to nullptr
+
```
Unnecessary empty line.

SeanSweeney04 replied:
Removed

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,156 @@
+}
+
+// Default constructor for LinkedList: Initializes head and tail as nullptr
+
```
Unnecessary empty line. Please fix it everywhere below.

SeanSweeney04 replied:
Fixed

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,156 @@
+        head = temp;
+        return;
+    }
+    else {
```
Egyptian style. Please correct it everywhere below.

SeanSweeney04 replied:
Fixed

gregory-sokolov replied:
Not fixed in this line.

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,30 @@
+
+    void clear();
+    Logger& _logger;
+    friend class LinkedListLogger;
```
Using friends are not the best design. Please rework it without using friend classes.

gregory-sokolov replied:
General comment: please make appropriate title and description for the Pull Request.

SeanSweeney04 replied:
Used getters and setters instead

gregory-sokolov commented:
General comment: please make the appropriate good-looking title and description for the PR.

Commit:
Used getters and setters for the Node class instead of the friend class. Also removed unecessary empty lines

SeanSweeney04 committed:
Merge branch 'main' into dev

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,163 @@
+    return head;
+}
+
+//Getter for tail
```
Missing space after //

SeanSweeney04 replied:
fixed

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,163 @@
+        insert(temp->data);
+        temp = temp->next;
+    }
+    logger.log(LogLevel::INFO, "Copy Constructor called.");
```
I would separate the final debugging line from the code with an empty line.

SeanSweeney04 replied:
done

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,163 @@
+{
+    if (this == &list) 
+    {
+        return *this;  // Avoid self-assignment
```
2 "return *this", here and in the end, may look duplicating and confusing because a reader expects different returns on failed condition and on success. It's better to follow the standard scheme of operator assignment implementation, many people simply get used to it.

SeanSweeney04 replied:
Done

gregory-sokolov replied:
Still not resolved.

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,163 @@
+                tail = prev; // Update tail if needed
+            }
+            delete cur;
+            return;
```
Return-from-the-middle, especially somewhere from the loop, is not the best programming style. Can you rewrite it with "break" for example?

SeanSweeney04 replied:
replaced with break

gregory-sokolov replied:
Still not resolved.

SeanSweeney04 replied:
Sorry, resolved now

gregory-sokolov commented on LinkedList/LinkedListLogger.h:
```diff
@@ -0,0 +1,18 @@
+#define LINKEDLISTLOGGER_H
+
+#include "LinkedList.h"
+#include <iostream>
```
Library first, your include next.

gregory-sokolov replied:
Still not resolved.

SeanSweeney04 replied:
Fixed

gregory-sokolov commented on LinkedList/Logger.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+
+//Destructor: Closes the log file
+
```
Extra empty line.

gregory-sokolov commented on LinkedList/Logger.cpp:
```diff
@@ -0,0 +1,61 @@
+}
+
+// Enables logging to a file
+
```
Extra empty line.

SeanSweeney04 replied:
fixed

gregory-sokolov commented on LinkedList/Logger.cpp:
```diff
@@ -0,0 +1,61 @@
+    case LogLevel::ERROR:
+        logLevelStr = "[ERROR] ";
+        break;
+    }
```
When you use "switch", there must be a default case (even empty with just a break).

SeanSweeney04 replied:
added default case

gregory-sokolov commented on LinkedList/Logger.cpp:
```diff
@@ -0,0 +1,61 @@
+    if (fileLoggingEnabled && logFile.is_open())
+    {
+        logFile << fullMessage << std::endl;
+    }
```
Liked that you combined both file and console logging in one place!

gregory-sokolov commented on LinkedList/Logger.cpp:
```diff
@@ -0,0 +1,61 @@
+{
+
+    logFile.open(filename, std::ios::app);
+    if (logFile.is_open()) {
```
One more Egyptian.

SeanSweeney04 replied:
fixed

gregory-sokolov commented on LinkedList/Logger.h:
```diff
@@ -0,0 +1,25 @@
+
+    void log(LogLevel level, const std::string& message);
+    void enableFileLogging(const std::string& filename);
+
```
Extra empty line.

SeanSweeney04 replied:
fixed

Commit:
Fixed 2 return this statements issue and added a default case for switch case

Commit:
Merge branch 'dev' of https://github.com/SeanSweeney04/LinkedList into dev

gregory-sokolov commented on LinkedList/main.cpp:
```diff
@@ -0,0 +1,46 @@
+        l1.insert(i);
+    }
+
+    std::cout << "Original List (l1): ";
```
I see you only testing some removals from the list. How about inserts? 
Also, more important was to add some object creation tests, like copying, assignment, passing to and out of function (by value, by ref and pointer). Need to extend the test case set.

SeanSweeney04 replied:
Done

gregory-sokolov commented on LinkedList/linkedlist_log.txt:
```diff
@@ -0,0 +1,136 @@
+[INFO] Program started.
```
I look at this log and it's not easy to read. Please separate test cases with empty lines or make a special log line like "Test 01: ...", "Test 02: ...". So far they all go in one continuous piece.

SeanSweeney04 replied:
Done

Commit:
added more test cases and made it more readable

Commit:
Fixed typo in logger message

gregory-sokolov commented on LinkedList/main.cpp:
```diff
@@ -0,0 +1,121 @@
+    l3Logger.print();
+    cout << "-----------------------------------------\n";
+
+    // ✅ Test 06: Passing by Reference (formerly passing by value)
```
The test says it's passing by value, but actually it's passing by reference. What's the reason of having 2 same test cases?

SeanSweeney04 replied:
Fixed

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,33 @@
+#include <cstddef> 
+#include "Node.h"
+#include "Logger.h"
+#include "LinkedListLogger.h"
```
Do you remember, we discussed that there should be minimum of dependencies in the header file? List obviously shouldn't depend on logging facilities. See also the comment below.

SeanSweeney04 replied:
Fixed

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,33 @@
+    Node* tail = nullptr;
+
+    void clear();
+    Logger& _logger;
```
As we discussed previously, Logger should not be the part of the Data class. It breaks SRP. If you want to use it in ctors/dtors, you need some other way to do it, not via member field.

SeanSweeney04 replied:
Fixed

gregory-sokolov commented:
General comment. Still need a good PR title and description because "New pull request after restoring the file structure" is not what your work was about. There are some unresolved conversations from previous review iterations, please check the conversations are resolved. Also, please add the latest log file from your program run because the current one is messy.

Commit:
Linked List class no longer implements the logger, abiding to the SRP

SeanSweeney04 commented:
I can't see an option for adding a description to the PR, only a title

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,28 @@
+
+#include <cstddef> 
+#include "Node.h"
+#include <iostream>
```
Still not the best design solution. There shouldn't be stream and cstd library dependencies in your header file. Just remove those includes and see what you can do.

SeanSweeney04 replied:
Removed

gregory-sokolov replied:
Still not resolved, iostream is there.

gregory-sokolov commented on LinkedList/LinkedList.h:
```diff
@@ -0,0 +1,28 @@
+    ~LinkedList();
+
+    void insert(int x);
+    void remove(int x, std::ostream& logStream);
```
The same comment: get rid of STL stream usage in the header. LinkedList interface must know nothing about streams. However, it's OK to use streams and other libs in cpp.

SeanSweeney04 replied:
Now logging happens inside LinkedList.cpp, not in LinkedList.h.

gregory-sokolov commented on LinkedList/linkedlist_log.txt:
```diff
@@ -0,0 +1,14 @@
+[INFO] TEST SUITE STARTED
+[INFO] Test 01: Creating and inserting elements into l1.
```
I don't see any ctor/dtor logs. They were in the previous version and you need them back because the whole reason of this task was to explore when and how they are called.
It's better to separate test cases by empty line.

I'm expecting to see something like this:
Test Case 01: name of test
test-related logging
ctor/dtor debugging

Test Case 02: name of test
...

SeanSweeney04 replied:
added

Commit:
All logging is now done outside of LinkedList interface, also added ctor/dtor loggs in main

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,154 @@
+        temp = temp->next;
+    }
+
+    std::cout << "[DEBUG] LinkedList assignment operator called." << std::endl;
```
You have a good logger with debug levels, use it. Creating separate debug outputs with cout/cerr looks duplicating and unnecessary, also you won't see those messages in the file.

SeanSweeney04 replied:
Done

Commit:
Removed second retrun this, also added debug logs to main

Commit:
Created Temp class

SeanSweeney04 commented:
Should the loggers be called automatically when a constructor is called or should this only happen in the main function? I have both in right now, I'll remove whichever one you think is not needed

Commit:
Refactored Logger to use static design without singleton. If this you're happy with this design, I'll add more logs to the main for caliing ctors/dtors

- Made Logger::log() a static method for global access
- Added Logger::initialize() to open the log file once at startup
- Removed use of singleton pattern (getLog()) as per reviewer guidance
- Introduced Logger::shutdown() to flush and close the log file at program end
- Ensured consistent log output to both console and file

gregory-sokolov commented:
How to edit title and description:
![edit-pr](https://github.com/user-attachments/assets/86980232-adbc-4f87-ab92-3344d5d3067f)

gregory-sokolov commented on LinkedList/LinkedListLogger.cpp:
```diff
@@ -0,0 +1,28 @@
+std::string LinkedListLogger::toString() const
+{
+    if (!list)
+        return "[ERROR] LinkedListLogger is referencing a null LinkedList.";
```
Even if there is one line in if-else-while-for, it should be placed in {}

SeanSweeney04 replied:
fixed

gregory-sokolov commented on LinkedList/LinkedListLogger.cpp:
```diff
@@ -0,0 +1,28 @@
+
+    Node* temp = list->getHead();
+    if (!temp)
+        return "List is empty.";
```
Even if there is one line in if-else-while-for, it should be placed in {}

SeanSweeney04 replied:
fixed

gregory-sokolov commented on LinkedList/Temp.cpp:
```diff
@@ -0,0 +1,27 @@
+
+Temp::Temp()
+{
+    lists.emplace_back();  // Stack
```
Until we studied move semantics, I recommend not to use it at all. Using simple push_back or vector initializer { ... } is more than enough.
The same below with std:: move - not needed.

SeanSweeney04 replied:
Done

gregory-sokolov commented on LinkedList/main.cpp:
```diff
@@ -0,0 +1,122 @@
+    for (auto& list : temp.getLists())
+    {
+        list.insert(value++);
+        list.insert(value++);
```
The idea of Temp class was not to test linked list functionality but to test copying and destruction in a more complex scenario. That's why you need to pass it to some function by value or do some similar actions, involving copying.

SeanSweeney04 replied:
Changed

gregory-sokolov reviewed (changes requested):
Logger design is ok, you can add more logs.

Commit:
re factored Temp class

gregory-sokolov commented on LinkedList/linkedlist_log.txt:
```diff
@@ -0,0 +1,63 @@
+[INFO] Original l1 after pass-by-pointer:
+[INFO] List contents: 1 3 100 101 
+[INFO] -----------------------------------------
+[INFO] ================== Temp Test ==================
```
Temp test is just Test 09, please keep the same good test header as for other tests.

SeanSweeney04 replied:
Done

gregory-sokolov commented on LinkedList/LinkedList.cpp:
```diff
@@ -0,0 +1,138 @@
+}
+
+// Copy Constructor: Creates a deep copy of another LinkedList
+LinkedList::LinkedList(const LinkedList& list)
```
The logging is missing for all ctors/dtors. We wanted to explore their calls, right? Please add them.

SeanSweeney04 replied:
Done

Commit:
Added logs to constructors instead of calling them manually in main.

Commit:
Changed constructor logs from info to debug

gregory-sokolov commented on LinkedList/linkedlist_log.txt:
```diff
@@ -0,0 +1,141 @@
+[INFO] =========================================
+[INFO] TEST SUITE ENDED
+[INFO] =========================================
+[DEBUG] LinkedList destructor called.
```
Trailing destructor logging after the "Test Suite Ended" message doesn't look very nice. End message should be the last in the log. Maybe it's better to put each test into a scope block {} which is also useful in terms of construction/destruction checks?

SeanSweeney04 replied:
Done

gregory-sokolov replied:
Still not resolved. Tests 1-5 and 9 not placed into {}.
I would recommend put the whole test between: 
[INFO] Test XX: Passing Temp instance by value and reference
[INFO] -----------------------------------------
into the scope {}.

SeanSweeney04 replied:
They should be in the scope now

gregory-sokolov replied:
Please attach a new log with corrected output.

SeanSweeney04 replied:
What is the corrected output in this case?

gregory-sokolov replied:
Oh, I see you have already added it to the latest commit.
Then the initial problem still remains: there are some destructor logging outside of test case boundaries. 

[INFO] -----------------------------------------
--> The last test has ended (the separator above), but there is some destructor logging below. Such output may confuse a user/test engineer in real applications, because it looks like you forgot to deallocate resources within the test. I suggest returning separate scopes for each test both with keeping your existing general scope.
[INFO] 
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[DEBUG] LinkedList destructor called.
[INFO] =========================================
[INFO] TEST SUITE ENDED
[INFO] =========================================

SeanSweeney04 replied:
So like this for example? : Logger::initialize("linkedlist_log.txt");
{
    Logger::log(INFO, "=========================================");
    Logger::log(INFO, "TEST SUITE STARTED");
    Logger::log(INFO, "=========================================");
    {
        Logger::log(INFO, "Test 01: Creating and inserting elements into l1.");
        LinkedList l1;
        LinkedListLogger l1Logger(l1);

        for (int i = 0; i < 5; i++)
        {
            l1.insert(i);
        }

        Logger::log(INFO, "Original List (l1): " + l1Logger.toString());
        Logger::log(INFO, "-----------------------------------------");
        Logger::log(LogLevel::INFO, "");
    }

gregory-sokolov replied:
Yes, exactly.

SeanSweeney04 replied:
Done

gregory-sokolov commented on LinkedList/Temp.h:
```diff
@@ -0,0 +1,20 @@
+    Temp(const Temp& other);
+    Temp& operator=(const Temp& other);
+
+    std::vector<LinkedList>& getLists();
```
Could be made const.

SeanSweeney04 replied:
Changed to const

Commit:
wrapped each test in a scope block

gregory-sokolov commented on LinkedList/main.cpp:
```diff
@@ -0,0 +1,135 @@
+        Logger::log(INFO, "-----------------------------------------");
+
+        Logger::log(INFO, "Test 09: Passing Temp instance by value and reference");
+
```
The last test also should be places into {}, like all other above.

SeanSweeney04 replied:
Done

gregory-sokolov commented on LinkedList/main.cpp:
```diff
@@ -0,0 +1,135 @@
+#include "LinkedListLogger.h"
+#include "Logger.h"
+#include "Temp.h"
+#include <iostream>
```
Library includes go first, your includes go next. It's better to separate them with an empty line.

SeanSweeney04 replied:
Done

gregory-sokolov commented on LinkedList/main.cpp:
```diff
@@ -0,0 +1,135 @@
+    Logger::log(DEBUG, "Leaving passTemp() with a return by value");
+
+    return temp_val;
+
```
Extra empty line.

SeanSweeney04 replied:
removed

gregory-sokolov commented on LinkedList/linkedlist_log.txt:
```diff
@@ -0,0 +1,141 @@
+[INFO] Original l1 after pass-by-pointer:
+[INFO] List contents: 1 3 100 101 
+[INFO] -----------------------------------------
+[INFO] Test 09: Passing Temp instance by value and reference
```
It would be much easier to read the log if separate each test case with an empty line.

SeanSweeney04 replied:
Done

gregory-sokolov commented on LinkedList/LinkedListLogger.cpp:
```diff
@@ -0,0 +1,32 @@
+#include "Logger.h"
+#include <iostream>
+
+
```
Extra empty line.

SeanSweeney04 replied:
Removed

Commit:
Fixed main scope

Commit:
Placed each test in a separate scope
